### ISSUE TITLE
Test Flake: Unit test cmd/entrypoint TestRealRunnerTimeout

### ISSUE THREAD

**Issue by lbernick** [tektoncd/pipeline#4643]
Sadly not much info present in log file:
```
=== RUN   TestRealRunnerTimeout
    runner_test.go:37: step didn't timeout
--- FAIL: TestRealRunnerTimeout (0.03s)
```

/kind flake
/priority important-soon

**Comment by abayer** [comment]
/lifecycle frozen

I hit this one while trying to reproduce the various events-related flakes, so it's still a thing. I thought that maybe increasing the timeout in the test would help, but I seem to still be hitting it...

**Comment by abayer** [comment]
Yeah, it failing or not depended on what I had for `-count` and the overall load on my system, not on anything flaky. Sigh. Well, it's still a flake, just one I can't come up with a fix for...

**Comment by Yongxuanzhang** [comment]
I'm looking into this issue and I'm not sure what the testcase wants to test. From the test description it wants to test that the runner will be killed after a millisecond (though it wants to sleep 10 ms?),
https://github.com/tektoncd/pipeline/blob/0b8349b770a76877051c9c790c94bf9ed897c75c/cmd/entrypoint/runner_test.go#L123
and the DeadlineExceeded error can be retuned in 2 places:
https://github.com/tektoncd/pipeline/blob/f83cd1f63bb216c76a3af46d58614d517113d1c5/cmd/entrypoint/runner.go#L120-L125
https://github.com/tektoncd/pipeline/blob/f83cd1f63bb216c76a3af46d58614d517113d1c5/cmd/entrypoint/runner.go#L154-L158
It seems that the DeadlineExceeded can still be returned if the runner really sleep 10 ms?

**Comment by lbernick** [comment]
> I'm looking into this issue and I'm not sure what the testcase wants to test. From the test description it wants to test that the runner will be killed after a millisecond (though it wants to sleep 10 ms?),

The command here that the entrypoint is supposed to be running is "sleep 10ms". This test ensures that the runner can be timed out in the appropriate time. i.e. it wants to ensure that the runner times out after 1ms even though the command is not done.

> It seems that the DeadlineExceeded can still be returned if the runner really sleep 10 ms?

Maybe? If so, that would mean that our tests could pass because the runner finished running and not because it got timed out correctly, which isn't great. (I suspect our timeout logic works correctly but ideally our tests should tell us this!) To test out whether or not this could happen, you could set the timeout to a very long time, and see if the tests still pass when the "sleep" command completes after 10ms.

**Comment by Yongxuanzhang** [comment]
From the test and code, it uses `context.WithTimeout(context.Background(), timeout)` this timeout ctx to count the time, and the `ctx` is also added to the `cmd`, right now the error is returned here:
https://github.com/tektoncd/pipeline/blob/09c7285609717d72a318058806b18fc6da4e9643/cmd/entrypoint/runner.go#L120
And if we jump into the `Start()`, it will return before the process is created to execute the cmd b/c the ctx is already done with timeout err
<img width="690" alt="image" src="https://user-images.githubusercontent.com/44033547/205353884-523f642a-e142-4b39-82ea-b0bf1dec806c.png">
